cloud-manager/pkg/services/kubernetes_service.go

```diff
@@ -0,0 +1,181 @@
+		IgnoreDaemonsets:   ignoreDaemonsets,
+		GracePeriodSeconds: gracePeriodSeconds,
+		Timeout:            timeout * time.Second,
+		DeleteLocalData:    deleteLocalData,
+		Namespace:          namespace,
+		Logger:             s,
+	}
+
+	for _, item := range nodeList.Items {
+		node := &item
+		nodes := []*corev1.Node{node}
+
+		log.Println(fmt.Sprintf("Draining node %s", node.GetName()))
+		err = k8sdrain.Drain(s.client, nodes, drainOptions)
+		if err != nil {
+			log.Println(fmt.Sprintf("Failed to drain node %s due to %s", node.GetName(), err.Error()))
+			return err
+		}
+
+		// teardown ec2 instance
+		log.Println(fmt.Sprintf("Deleting node: %s", node.GetName()))
+		isTerminated, err := s.provider.TerminateInstance(node.GetName(), forceTermination)
+		if err != nil {
+			err := errors.New(fmt.Sprintf("Failed to terminate node %s due to %s", node.GetName(), err.Error()))
+			return err
+		}
+		if !isTerminated {
+			err := errors.New("Termination failed, please check if instance allows termination and/or specify the flag -force-termination")
+			return err
+		}
+
+		isReady, err := s.verifyNew(initialNodeSize)
+		if err != nil {
+			return err
+		}
+		if !isReady {
+			return errors.New("Node is not ready")
+		}
+	}
+	return nil
+}
+
+// TODO: Does not handle race conditions from async creation of nodes when the script is running
+func (s *Service) verifyNew(initialSize int) (bool, error) {
+	nodeList, err := s.client.CoreV1().Nodes().List(metav1.ListOptions{})
+	if err != nil {
+		return false, err
+	}
+
+	readyNodes := s.filterReadyNodes(nodeList.Items)
+
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Minute)
+	defer cancel()
+
+	for {
+		select {
+		case <-ctx.Done():
+			err := errors.New("Timeout reached, while waiting for nodes to be ready")
+			return false, err
+		default:
+			log.Println(fmt.Sprintf("Nodes ready: %x of %x", len(readyNodes), initialSize))
+			if len(readyNodes) == initialSize {
+				return true, nil
+			}
+			time.Sleep(15 * time.Second)
+			nodeList, err := s.client.CoreV1().Nodes().List(metav1.ListOptions{})
+
+			if err != nil {
+				return false, err
+			}
+			readyNodes = s.filterReadyNodes(nodeList.Items)
+
+		}
+	}
+}
+
+func (s *Service) Log(v ...interface{}) {
+	log.Println(v)
+}
+
+func (s *Service) Logf(format string, v ...interface{}) {
+	log.Println(fmt.Sprintf(format, v))
+}
+
+// check if kubelet status is ready
+func (s *Service) nodeIsReady(node corev1.Node) bool {
+	for _, condition := range node.Status.Conditions {
+		if condition.Type == "Ready" && condition.Reason == "KubeletReady" {
+			return s.nodeIsSchedulable(node.Spec.Taints)
+		}
+	}
+	return false
+}
+
+// check if node has any taints and if true if any taint is for unschedulable
+func (s *Service) nodeIsSchedulable(taints []corev1.Taint) bool {
+	if len(taints) == 0 {
+		return true
+	} else {
+		for _, taint := range taints {
+			if taint.Key != PodUnschedulable {
+				return true
+			}
+		}
```

Comment:
This would possibly return true when the slice of taints gets checked. If the first one != PodUnschedulable then true is returned. The others taints have a key of PodUnschedulable then they won't get checked.

Reply:
Thanks @gabrieljackson I have updated it with a sum checker!